### PR TITLE
[feature] Added retry mechanism to SeleniumTestMixin #464

### LINKED ISSUE #464: [tests] Retry failing selenium tests
We need a mechanism which would automatically retry failing selenium tests a couple of times before flagging them as failed.

With the help of ChatGPT I generated a prototype for this:

```python
import unittest
import time
import traceback
import sys


class SeleniumTestMixin:
    retry_max = 2
    retry_delay = 0  # seconds

    def run(self, result=None):
        """
        Override unittest.TestCase.run to retry flaky tests.
        """
        original_result = result
        test_name = self.id()

        for attempt in range(self.retry_max + 1):
            temp_result = self.defaultTestResult()
            super().run(temp_result)

            if temp_result.wasSuccessful():
                if original_result is not None:
                    original_result.addSuccess(self)
                return

            if attempt < self.retry_max:
                print("-" * 80)
                print(
                    f"[Retry] Test '{test_name}' failed on attempt {attempt + 1}. Retrying..."
                )
                print("-" * 80)
                if self.retry_delay:
                    time.sleep(self.retry_delay)
                continue
            else:
                for failed_test, err in temp_result.failures + temp_result.errors:
                    print("-" * 80)
                    print(f"[Traceback for '{failed_test.id()}']:\n")
                    print(err.strip())
                    print("-" * 80)

                if original_result is not None:
                    for failure in temp_result.failures:
                        original_result.addFailure(*failure)
                    for error in temp_result.errors:
                        original_result.addError(*error)
```

We need to update the docs to state that selenium tests are automatically retried and document `retry_max` and `retry_delay`, plus do some manual testing to ensure that the messages printed when flaky tests are retried are visible and readable.

### PR DESCRIPTION
**Pull request by pandafy**
## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [x] I have updated the documentation.

## Reference to Existing Issue

Closes #464

## Description of Changes

Retry selenium tests if the tests fails on the first attempt. This prevents failng the CI build from flaky tests.

### CODE REVIEW

**Comment by pandafy** [comment]
The test suite would fail if the test is flaky 

```
Found 1 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_fails", attempt 1/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_fails", attempt 2/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_fails", attempt 3/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_fails", attempt 4/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_fails", attempt 5/5. 
--------------------------------------------------------------------------------

Summary of slow tests (>0.3s)


Total slow tests detected: 0

======================================================================
FAIL: test_retry_mechanism_fails (test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pandafy/openwisp/openwisp-utils/tests/test_project/tests/test_selenium.py", line 796, in test_retry_mechanism_fails
    self.fail('Report failed tests')
AssertionError: Report failed tests

----------------------------------------------------------------------
Ran 1 test in 2.187s

FAILED (failures=1)
Destroying test database for alias 'default'...
```

The test will be considered as pass, it it has success rate of 80%

```
Found 1 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_pass", attempt 1/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_pass", attempt 2/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_pass", attempt 3/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_pass", attempt 4/5. 
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
[Retry] Retrying "test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism.test_retry_mechanism_pass", attempt 5/5. 
--------------------------------------------------------------------------------
.
Summary of slow tests (>0.3s)

(test_project.tests.test_selenium.TestSeleniumMixinRetryMechanism)
  (2.15s) test_retry_mechanism_pass

Total slow tests detected: 1

----------------------------------------------------------------------
Ran 1 test in 2.150s

OK
Destroying test database for alias 'default'...
````

File: docs/developer/test-utilities.rst [hunk 1]
```diff
@@ -179,6 +179,35 @@ This mixin provides the core Selenium setup logic and reusable test
 methods that must be used across all OpenWISP modules based on Django to
 enforce best practices and avoid flaky tests.
 
+It also provides a retry mechanism for flaky tests, which executes the a
+failing test multiple times to determine if it is a transient failure.
+This functionality can be configured using the following class attributes:
+
+- ``retry_max``: Number of times to retry a test (default: ``5``).
+- ``retry_delay``: Numbed of seconds to wait before retrying a test
+  (default: ``2``).
```
Comment (nemesifier) [inline docs/developer/test-utilities.rst:188]: Why do we need to wait to repeat tests? I'd set this to zero to avoid increasing build times.

File: tests/test_project/tests/test_selenium.py [hunk 2]
```diff
@@ -753,3 +753,39 @@ def test_get_browser_logs(self):
...
+        pass
+
+    @classmethod
+    def tearDownClass(cls):
+        pass
+
+    def test_retry_mechanism_pass(self):
+        if not hasattr(self, '_test_retry_mechanism_pass_called'):
+            self._test_retry_mechanism_pass_called = 1
+            self.fail("Failing on first call")
```
Comment (nemesifier) [inline tests/test_project/tests/test_selenium.py:773]:
```suggestion
            self.fail('Failing on first call')
```

**Review by nemesifier — CHANGES_REQUESTED** [review]
Please double check the docs text.

**Review by nemesifier — APPROVED** [review]
